src/lib/bootloader/proposal_client.rb

```diff
@@ -7,6 +7,38 @@
 module Bootloader
   # Proposal client for bootloader configuration
   class ProposalClient < ::Installation::ProposalClient
+    # Error when during update media is booted by different technology than target system.
+    class MismatchBootloader < RuntimeError
+      include Yast::I18n
+
+      def initialize(old_bootloader, new_bootloader)
+        @old_bootloader = old_bootloader
+        @new_bootloader = new_bootloader
+
+        super("Mismatching bootloaders")
+      end
+
+      def user_message
+        textdomain "bootloader"
+
+        boot_map = {
+          # TRANSLATORS: kind of boot. It is term for way how x86_64 can boot
+          "grub2"     => _("Legacy BIOS boot"),
+          # TRANSLATORS: kind of boot. It is term for way how x86_64 can boot
+          "grub2-efi" => _("EFI boot")
+        }
+        # TRANSLATORS: keep %{} intact. It will be replaced by kind of boot
+        format(_(
+                 "Cannot update bootloader as there is mismatch of boot technology.<br>" \
+                   "System to update is booted via %{old_boot} and " \
+                   "media is booted via %{new_boot}.<br>" \
+                   "This scenario is not supported. Updated system may stop booting " \
+                   "when continue or update itself can fail."
+        ),
+          old_boot: boot_map[@old_bootloader], new_boot: boot_map[@new_bootloader])
```

Comment:
What if `@old_bootloader` or `@new_bootloader` value is missing in the `boot_map`? That will result in `nil` and empty strings. To make it robust I'd suggest to use the original values without mapping. The class name `MismatchBootloader` suggests it can handle any bootloaders, not only grub2.

Or alternatively rename it to something like `MismatchGrub2Bootloader` so it is clear it's designed only for grub2.

Reply:
but then it will be less useful for users to see grub2 and grub2-efi instead of that legacy BIOS and EFI. I can ensure it in constructor, that it raises if something else is passed. Is it ok for you?

Reply:
Yes, that's OK.